File: flutter-app/lib/services/history_manager.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

/// クイズの結果を保存するためのクラス
class QuizHistory {
  final String id;
  final String genre;
  final String responderName;
  final int score;
  final int total;
  final int timeMillis;
  final DateTime timestamp;
  final List<QuestionResult> questionResults;

  QuizHistory({
    required this.id,
    required this.genre,
    required this.responderName,
    required this.score,
    required this.total,
    required this.timeMillis,
    required this.timestamp,
    required this.questionResults,
  });

  double get accuracy => total > 0 ? score / total * 100 : 0;

  Map<String, dynamic> toJson() => {
    'id': id,
    'genre': genre,
    'responderName': responderName,
    'score': score,
    'total': total,
    'timeMillis': timeMillis,
    'timestamp': timestamp.toIso8601String(),
    'questionResults': questionResults.map((q) => q.toJson()).toList(),
  };

  factory QuizHistory.fromJson(Map<String, dynamic> json) => QuizHistory(
    id: json['id'],
    genre: json['genre'],
    responderName: json['responderName'] ?? '',
    score: json['score'],
    total: json['total'],
    timeMillis: json['timeMillis'],
    timestamp: DateTime.parse(json['timestamp']),
    questionResults: (json['questionResults'] as List?)
        ?.map((q) => QuestionResult.fromJson(q))
        .toList() ?? [],
  );
}

/// 各問題の結果
class QuestionResult {
  final int questionNumber;
  final String description;
  final bool isCorrect;
  final bool wasSame; // 正解が「同じ」だったか
  final bool answeredSame; // ユーザーの回答が「同じ」だったか

  QuestionResult({
    required this.questionNumber,
    required this.description,
    required this.isCorrect,
    required this.wasSame,
    required this.answeredSame,
  });

  Map<String, dynamic> toJson() => {
    'questionNumber': questionNumber,
    'description': description,
    'isCorrect': isCorrect,
    'wasSame': wasSame,
    'answeredSame': answeredSame,
  };

  factory QuestionResult.fromJson(Map<String, dynamic> json) => QuestionResult(
    questionNumber: json['questionNumber'],
    description: json['description'],
    isCorrect: json['isCorrect'],
    wasSame: json['wasSame'],
    answeredSame: json['answeredSame'],
  );
}

/// 履歴管理クラス
class HistoryManager {
  static const String _key = 'quiz_history';
  static HistoryManager? _instance;
  
  List<QuizHistory> _histories = [];
  
  static HistoryManager get instance {
    _instance ??= HistoryManager._();
    return _instance!;
  }
  
  HistoryManager._();
  
  List<QuizHistory> get histories => List.unmodifiable(_histories);
  
  /// 履歴を読み込む
  Future<void> loadHistories() async {
    final prefs = await SharedPreferences.getInstance();
    final jsonStr = prefs.getString(_key);
    if (jsonStr != null) {
      final List<dynamic> jsonList = jsonDecode(jsonStr);
      _histories = jsonList.map((j) => QuizHistory.fromJson(j)).toList();
      // 新しい順にソート
      _histories.sort((a, b) => b.timestamp.compareTo(a.timestamp));
    }
  }
  
  /// 履歴を保存
  Future<void> saveHistory(QuizHistory history) async {
    _histories.insert(0, history);
    await _persist();
  }
  
  /// 永続化
  Future<void> _persist() async {
    final prefs = await SharedPreferences.getInstance();
    final jsonStr = jsonEncode(_histories.map((h) => h.toJson()).toList());
    await prefs.setString(_key, jsonStr);
  }
  
  /// 履歴をクリア
  Future<void> clearHistories() async {
    _histories.clear();
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_key);
  }
  
  /// 指定したIDの履歴を削除
  Future<void> deleteHistory(String id) async {
    _histories.removeWhere((h) => h.id == id);
    await _persist();
  }
  
  /// 指定した複数のIDの履歴を削除
  Future<void> deleteHistories(Set<String> ids) async {
    _histories.removeWhere((h) => ids.contains(h.id));
    await _persist();
  }
  
  /// ジャンルごとの統計
  Map<String, GenreStats> getStatsByGenre() {
    final Map<String, List<QuizHistory>> grouped = {};
    for (final h in _histories) {
      grouped.putIfAbsent(h.genre, () => []).add(h);
    }
    
    return grouped.map((genre, histories) => MapEntry(
      genre,
      GenreStats.fromHistories(genre, histories),
    ));
  }
  
  /// 全体の統計
  GenreStats getOverallStats() {
    return GenreStats.fromHistories('全体', _histories);
  }
  
  /// 回答者ごとの統計
  Map<String, GenreStats> getStatsByResponder() {
    final Map<String, List<QuizHistory>> grouped = {};
    for (final h in _histories) {
      final name = h.responderName.isEmpty ? '匿名' : h.responderName;
      grouped.putIfAbsent(name, () => []).add(h);
    }
    
    return grouped.map((name, histories) => MapEntry(
      name,
      GenreStats.fromHistories(name, histories),
    ));
  }
  
  /// ユニークな回答者リスト
  List<String> getResponders() {
    final set = <String>{};
    for (final h in _histories) {
      set.add(h.responderName.isEmpty ? '匿名' : h.responderName);
    }
    return set.toList()..sort();
  }
}

/// ジャンル/回答者ごとの統計
class GenreStats {
  final String name;
  final int totalTests;
  final int totalQuestions;
  final int totalCorrect;
  final int totalTimeMillis;
  final double averageAccuracy;
  final double averageScore;
  final double averageTime;

  GenreStats({
    required this.name,
    required this.totalTests,
    required this.totalQuestions,
    required this.totalCorrect,
    required this.totalTimeMillis,
    required this.averageAccuracy,
    required this.averageScore,
    required this.averageTime,
  });

  factory GenreStats.fromHistories(String name, List<QuizHistory> histories) {
    if (histories.isEmpty) {
      return GenreStats(
        name: name,
        totalTests: 0,
        totalQuestions: 0,
        totalCorrect: 0,
        totalTimeMillis: 0,
        averageAccuracy: 0,
        averageScore: 0,
        averageTime: 0,
      );
    }

    int totalQuestions = 0;
    int totalCorrect = 0;
    int totalTimeMillis = 0;

    for (final h in histories) {
      totalQuestions += h.total;
      totalCorrect += h.score;
      totalTimeMillis += h.timeMillis;
    }

    return GenreStats(
      name: name,
      totalTests: histories.length,
      totalQuestions: totalQuestions,
      totalCorrect: totalCorrect,
      totalTimeMillis: totalTimeMillis,
      averageAccuracy: totalQuestions > 0 ? totalCorrect / totalQuestions * 100 : 0,
      averageScore: totalCorrect / histories.length,
      averageTime: totalTimeMillis / histories.length,
    );
  }
}

File: android-app/app/src/main/java/com/tqmane/similarityquiz/HistoryManager.kt
package com.tqmane.similarityquiz

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONArray
import org.json.JSONObject
import java.io.Serializable

/**
 * 各問題の結果を保持するデータクラス（履歴用）
 */
data class QuestionResultData(
    val questionNumber: Int,
    val description: String,
    val isCorrect: Boolean,
    val wasSame: Boolean,
    val answeredSame: Boolean
) : Serializable {
    fun toJson(): JSONObject = JSONObject().apply {
        put("questionNumber", questionNumber)
        put("description", description)
        put("isCorrect", isCorrect)
        put("wasSame", wasSame)
        put("answeredSame", answeredSame)
    }

    companion object {
        fun fromJson(json: JSONObject): QuestionResultData = QuestionResultData(
            questionNumber = json.getInt("questionNumber"),
            description = json.optString("description", ""),
            isCorrect = json.getBoolean("isCorrect"),
            wasSame = json.getBoolean("wasSame"),
            answeredSame = json.getBoolean("answeredSame")
        )
    }
}

/**
 * クイズ履歴を保持するデータクラス
 */
data class QuizHistoryData(
    val id: String,
    val genre: String,
    val responderName: String,
    val score: Int,
    val total: Int,
    val timeMillis: Long,
    val timestamp: Long,
    val questionResults: List<QuestionResultData>
) : Serializable {
    val accuracy: Double get() = if (total > 0) score.toDouble() / total * 100 else 0.0

    fun toJson(): JSONObject = JSONObject().apply {
        put("id", id)
        put("genre", genre)
        put("responderName", responderName)
        put("score", score)
        put("total", total)
        put("timeMillis", timeMillis)
        put("timestamp", timestamp)
        put("questionResults", JSONArray().apply {
            questionResults.forEach { put(it.toJson()) }
        })
    }

    companion object {
        fun fromJson(json: JSONObject): QuizHistoryData = QuizHistoryData(
            id = json.getString("id"),
            genre = json.getString("genre"),
            responderName = json.optString("responderName", ""),
            score = json.getInt("score"),
            total = json.getInt("total"),
            timeMillis = json.getLong("timeMillis"),
            timestamp = json.getLong("timestamp"),
            questionResults = json.optJSONArray("questionResults")?.let { arr ->
                (0 until arr.length()).map { QuestionResultData.fromJson(arr.getJSONObject(it)) }
            } ?: emptyList()
        )
    }
}

/**
 * ジャンル/回答者ごとの統計
 */
data class GenreStats(
    val name: String,
    val totalTests: Int,
    val totalQuestions: Int,
    val totalCorrect: Int,
    val totalTimeMillis: Long,
    val averageAccuracy: Double,
    val averageScore: Double,
    val averageTime: Double
) {
    companion object {
        fun fromHistories(name: String, histories: List<QuizHistoryData>): GenreStats {
            if (histories.isEmpty()) {
                return GenreStats(name, 0, 0, 0, 0, 0.0, 0.0, 0.0)
            }

            var totalQuestions = 0
            var totalCorrect = 0
            var totalTimeMillis = 0L

            for (h in histories) {
                totalQuestions += h.total
                totalCorrect += h.score
                totalTimeMillis += h.timeMillis
            }

            return GenreStats(
                name = name,
                totalTests = histories.size,
                totalQuestions = totalQuestions,
                totalCorrect = totalCorrect,
                totalTimeMillis = totalTimeMillis,
                averageAccuracy = if (totalQuestions > 0) totalCorrect.toDouble() / totalQuestions * 100 else 0.0,
                averageScore = totalCorrect.toDouble() / histories.size,
                averageTime = totalTimeMillis.toDouble() / histories.size
            )
        }
    }
}

/**
 * 履歴管理クラス
 */
class HistoryManager private constructor(context: Context) {
    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val histories: MutableList<QuizHistoryData> = mutableListOf()
    private val responderNames: MutableList<String> = mutableListOf()

    init {
        loadHistories()
        loadResponderNames()
    }

    private fun loadHistories() {
        val jsonStr = prefs.getString(KEY_HISTORIES, null) ?: return
        try {
            val arr = JSONArray(jsonStr)
            histories.clear()
            for (i in 0 until arr.length()) {
                histories.add(QuizHistoryData.fromJson(arr.getJSONObject(i)))
            }
            // 新しい順にソート
            histories.sortByDescending { it.timestamp }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
    
    private fun loadResponderNames() {
        val jsonStr = prefs.getString(KEY_RESPONDER_NAMES, null) ?: return
        try {
            val arr = JSONArray(jsonStr)
            responderNames.clear()
            for (i in 0 until arr.length()) {
                responderNames.add(arr.getString(i))
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun getHistories(): List<QuizHistoryData> = histories.toList()
    
    fun getRecentResponderNames(): List<String> = responderNames.toList()

    fun saveHistory(history: QuizHistoryData) {
        histories.add(0, history)
        
        // 回答者名を履歴に追加
        if (history.responderName.isNotEmpty()) {
            addResponderName(history.responderName)
        }
        
        persist()
    }
    
    /**
     * 回答者名を履歴に追加
     */
    fun addResponderName(name: String) {
        val trimmed = name.trim()
        if (trimmed.isEmpty()) return
        
        // 既に存在する場合は削除して先頭に追加
        responderNames.remove(trimmed)
        responderNames.add(0, trimmed)
        
        // 最大8件まで保持
        while (responderNames.size > MAX_RESPONDER_NAMES) {
            responderNames.removeAt(responderNames.lastIndex)
        }
        
        persistResponderNames()
    }
    
    /**
     * 回答者名を履歴から削除
     */
    fun removeResponderName(name: String) {
        responderNames.remove(name)
        persistResponderNames()
    }
    
    private fun persistResponderNames() {
        val arr = JSONArray()
        responderNames.forEach { arr.put(it) }
        prefs.edit().putString(KEY_RESPONDER_NAMES, arr.toString()).apply()
    }

    private fun persist() {
        val arr = JSONArray()
        histories.forEach { arr.put(it.toJson()) }
        prefs.edit().putString(KEY_HISTORIES, arr.toString()).apply()
    }

    fun clearHistories() {
        histories.clear()
        prefs.edit().remove(KEY_HISTORIES).apply()
    }

    /**
     * 指定したIDの履歴を削除
     */
    fun deleteHistory(id: String) {
        histories.removeAll { it.id == id }
        persist()
    }
    
    /**
     * 指定した複数のIDの履歴を削除
     */
    fun deleteHistories(ids: Set<String>) {
        histories.removeAll { it.id in ids }
        persist()
    }

    fun getStatsByGenre(): Map<String, GenreStats> {
        val grouped = histories.groupBy { it.genre }
        return grouped.mapValues { (genre, list) -> GenreStats.fromHistories(genre, list) }
    }

    fun getOverallStats(): GenreStats {
        return GenreStats.fromHistories("全体", histories)
    }

    fun getStatsByResponder(): Map<String, GenreStats> {
        val grouped = histories.groupBy { if (it.responderName.isEmpty()) "匿名" else it.responderName }
        return grouped.mapValues { (name, list) -> GenreStats.fromHistories(name, list) }
    }

    fun getResponders(): List<String> {
        return histories.map { if (it.responderName.isEmpty()) "匿名" else it.responderName }.distinct().sorted()
    }

    companion object {
        private const val PREFS_NAME = "quiz_history"
        private const val KEY_HISTORIES = "histories"
        private const val KEY_RESPONDER_NAMES = "responder_names"
        private const val MAX_RESPONDER_NAMES = 8

        @Volatile
        private var instance: HistoryManager? = null

        fun getInstance(context: Context): HistoryManager {
            return instance ?: synchronized(this) {
                instance ?: HistoryManager(context.applicationContext).also { instance = it }
            }
        }
    }
}
